File: Queue.cpp
#include "Queue.h"

void pushQueue(NodeQueue*& head, NodeQueue*& tail, std::string value) {
    NodeQueue* newNode = new NodeQueue;
    newNode->key = value;
    newNode->next = nullptr;
    if (tail != nullptr) {
        tail->next = newNode;
    }
    else {
        head = newNode;
    }
    tail = newNode;
}

void popQueue(NodeQueue*& head, NodeQueue*& tail) {
    if (head == nullptr) return;
    NodeQueue* toDelete = head;
    head = head->next;
    if (head == nullptr) tail = nullptr;
    delete toDelete;
}

void printQueue(NodeQueue* head) {
    std::cout << "Queue: ";
    while (head != nullptr) {
        std::cout << head->key << " ";
        head = head->next;
    }
    std::cout << std::endl;
}

File: golang/queue.go
package main

import "fmt"

type NodeQueue struct {
	key  string
	next *NodeQueue
}

func pushQueue(head **NodeQueue, tail **NodeQueue, value string) {
	n := &NodeQueue{key: value}
	if *tail != nil {
		(*tail).next = n
	} else {
		*head = n
	}
	*tail = n
}

func popQueue(head **NodeQueue, tail **NodeQueue) {
	if *head == nil {
		return
	}
	*head = (*head).next
	if *head == nil {
		*tail = nil
	}
}

func printQueue(head *NodeQueue) {
	p := head
	for p != nil {
		fmt.Print(p.key, " ")
		p = p.next
	}
	fmt.Println()
}
